File: csharp/forth/Forth.cs
using Sprache;

public static class Forth
{
    public abstract class ParseEntity
    {
    };

    public class Empty : ParseEntity
    {
        public Empty() { }
    }

    public class FunctionBody : ParseEntity
    {
        public Func<Stack<int>, int>? builtIn = null;
        public string? body = null;
        public Dictionary<string, FunctionBody>? symbolTable = null;

        public FunctionBody() { }
        public FunctionBody(string body, Dictionary<string, FunctionBody>? symbolTable)
        {
            this.body = body;
            this.symbolTable = symbolTable == null ? null : new Dictionary<string, FunctionBody>(symbolTable);
        }


        public FunctionBody(Func<Stack<int>, int> fn, Dictionary<string, FunctionBody>? symbolTable)
        {
            this.builtIn = fn;
            this.symbolTable = symbolTable == null ? null : new Dictionary<string, FunctionBody>(symbolTable);
        }
    }

    public class NumericConstant : ParseEntity
    {
        public int value { get; set; }

        public NumericConstant(string number)
        {
            if (!int.TryParse(number, out int value))
                value = 0;
            this.value = value;
        }
    }

    public class Identifier : ParseEntity
    {
        public string value { get; set; }

        public Identifier(string name)
        {
            if (name != "-")
            {
                if (!name.Any((c) => Char.IsDigit(c) == false && c != '-'))
                    throw new InvalidOperationException("Identifiers cannot be numbers");
            }
            value = name;
        }
    }

    public class WordDeclare : ParseEntity
    {
        public string name { get; set; }
        public string value { get; set; }

        public WordDeclare(string name, string value)
        {
            this.name = name;
            this.value = value;
        }
    }

    public static class Grammar
    {
        static char[] allowedIdentifierChars = {
            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k',
            'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v',
            'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G',
            'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
            'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '0', '1', '2',
            '3', '4', '5', '6', '7', '8', '9', '-', '_', '+', '-',
            '*', '/', '.'};

        static Parser<NumericConstant> PositiveNumber = Parse.Digit.AtLeastOnce().Text().Select(value => new NumericConstant(value));
        static Parser<NumericConstant> NegativeNumber = (
            from sign in Parse.Char('-')
            from digits in Parse.Digit.AtLeastOnce().Text()
            select new NumericConstant(sign.ToString() + digits));
        static Parser<NumericConstant> Number = PositiveNumber.XOr(NegativeNumber);
        static Parser<Empty> NonToken = Parse.WhiteSpace.Many().Return(new Empty()).Named("Empty");
        static Parser<Identifier> Add = Parse.Char('+').Token().Return(new Identifier("+")).Named("+");
        static Parser<Identifier> Subtract = (from ch in Parse.Char('-')
                                              from next in Parse.Not(Parse.Digit)
                                              select ch).Token().Return(new Identifier("-")).Named("-");
        static Parser<Identifier> Multiply = Parse.Char('*').Token().Return(new Identifier("*")).Named("*");
        static Parser<Identifier> Divide = Parse.Char('/').Token().Return(new Identifier("/")).Named("/");

        static Parser<Identifier> Word = (
            from _Id in Parse.Chars(allowedIdentifierChars).Many().Text()
            where _Id.Length > 0
            select new Identifier(_Id)).Named("IDENTIFIER");

        static Parser<WordDeclare> Declaration =
        (from colon in Parse.Char(':').Once().Text()
         from ws1 in Parse.Optional(Parse.WhiteSpace.AtLeastOnce())
         from name in Word
         from ws2 in Parse.Optional(Parse.WhiteSpace.AtLeastOnce())
         from body in Parse.CharExcept(c => c == ';', "End on Semicolon").Many().Text()
         from ws3 in Parse.Optional(Parse.WhiteSpace.AtLeastOnce())
         from semicolon in Parse.Char(';').Once().Text()
         select new WordDeclare(name.value, body.Trim())
        ).Named("WORD_DECLARE");
        static Parser<Identifier> mathExpression = Add.Or(Subtract).Or(Multiply).Or(Divide);
        static Parser<ParseEntity> expression = NonToken.Or<ParseEntity>(Number).Or<ParseEntity>(Declaration).Or<ParseEntity>(Word).Or<ParseEntity>(mathExpression);
        public static Parser<IEnumerable<ParseEntity>> expressions = expression.Many();
    }


    public static void ParseLine(string instruction, Stack<int> stack, Dictionary<string, FunctionBody> symbolTable)
    {
        foreach (var op in Grammar.expressions.Parse(instruction))
        {
            if (op == null || op.GetType() == typeof(Forth.Empty))
            {
                continue;
            }
            else if (op.GetType() == typeof(Forth.NumericConstant))
            {
                NumericConstant? c = op as NumericConstant;
                if (c == null) continue;
                stack.Push(c.value);
            }
            else if (op.GetType() == typeof(Forth.WordDeclare))
            {
                WordDeclare? word = op as Forth.WordDeclare;
                if (word == null)
                    continue;

                // In forth you can overwrite function.
                // They also only run against their original symbol table to stop infinite loops.                
                symbolTable[word.name.ToLowerInvariant()] = new FunctionBody(word.value, symbolTable);
            }
            else if (op.GetType() == typeof(Forth.Identifier))
            {
                Identifier? word = op as Forth.Identifier;
                if (word == null)
                    continue;

                if (symbolTable.ContainsKey(word.value.ToLowerInvariant()))
                {
                    FunctionBody body = symbolTable[word.value.ToLowerInvariant()];
                    if (body.builtIn != null)
                    {
                        int _ = body.builtIn(stack);
                    }
                    else if (body.body != null)
                    {
                        ParseLine(body.body, stack, body.symbolTable != null ? body.symbolTable : symbolTable);
                    }
                    else
                    {
                        throw new Exception($"No function body for \"{word.value}\".");
                    }
                }
                else
                {
                    throw new InvalidOperationException($"Word: \"{word.value}\" not found.");
                }
            }
            else
            {
                throw new Exception($"Unhandled grammar: {op.ToString()}");
            }
        }
    }
    public static string Evaluate(string[] instructions)
    {
        Stack<int> stack = new();
        // Initialize with the built-in functions.
        // They do not require a symbol table.
        Dictionary<string, FunctionBody> symbolTable = new()
        {
            { "dup", new FunctionBody(DUP, null) },
            { "drop", new FunctionBody(DROP, null) },
            { "swap", new FunctionBody(SWAP, null) },
            { "over", new FunctionBody(OVER, null) },
            { "+", new FunctionBody(ADD, null) },
            { "-", new FunctionBody(SUBTRACT, null) },
            { "*", new FunctionBody(MULTIPLY, null) },
            { "/", new FunctionBody(DIVIDE, null) },
            { ".", new FunctionBody(PRINT, null) },
        };

        foreach (string instruction in instructions)
        {
            ParseLine(instruction, stack, symbolTable);

        }
        return String.Join(' ', (from elem in stack.Reverse()
                                 select elem.ToString()).ToList<string>());
    }

    public static int ADD(Stack<int> stack)
    {
        int first;
        int second;

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        second = stack.Pop();

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        first = stack.Pop();

        int ret = first + second;
        stack.Push(ret);
        return ret;
    }

    public static int SUBTRACT(Stack<int> stack)
    {
        int first;
        int second;

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        second = stack.Pop();

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        first = stack.Pop();

        int ret = first - second;
        stack.Push(ret);
        return ret;
    }

    public static int MULTIPLY(Stack<int> stack)
    {
        int first;
        int second;

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        second = stack.Pop();

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        first = stack.Pop();

        int ret = first * second;
        stack.Push(ret);
        return ret;
    }

    public static int DIVIDE(Stack<int> stack)
    {
        int first;
        int second;

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        second = stack.Pop();

        if (stack.Count < 1)
            throw new InvalidOperationException("Math operation requires two operands.");

        first = stack.Pop();

        int ret = first / second;
        stack.Push(ret);
        return ret;
    }

    public static int DUP(Stack<int> stack)
    {
        if (stack.Count == 0)
        {
            throw new InvalidOperationException("Stack is empty, cannot duplicate.");
        }
        int duplicate = stack.Peek();
        stack.Push(duplicate);
        return duplicate;
    }

    public static int DROP(Stack<int> stack)
    {
        if (stack.Count == 0)
            throw new InvalidOperationException("Stack is empty, cannot drop.");
        return stack.Pop();
    }

    public static int SWAP(Stack<int> stack)
    {
        int first;
        int second;
        if (stack.Count < 1)
            throw new InvalidOperationException("Swap requires two operands.");

        second = stack.Pop();

        if (stack.Count < 1)
            throw new InvalidOperationException("Swap requires two operands.");

        first = stack.Pop();

        stack.Push(second);
        stack.Push(first);
        return first;
    }

    public static int OVER(Stack<int> stack)
    {
        int first;
        int second;

        if (stack.Count < 1)
            throw new InvalidOperationException("Over requires two operands.");

        second = stack.Pop();

        if (stack.Count < 1)
            throw new InvalidOperationException("Over requires two operands.");

        first = stack.Pop();

        stack.Push(first);
        stack.Push(second);
        stack.Push(first);

        return first;
    }
    
    public static int PRINT(Stack<int> stack)
    {
        if (stack.Count < 1)
            throw new InvalidOperationException("PRINT requires something be on the stack.");

        int first = stack.Pop();

        Console.WriteLine(first);

        return first;
    }
/*
    public static void Main(String[] args)
    {

        Stack<int> stack = new();
        // Initialize with the built-in functions.
        // They do not require a symbol table.
        Dictionary<string, FunctionBody> symbolTable = new()
        {
            { "dup", new FunctionBody(DUP, null) },
            { "drop", new FunctionBody(DROP, null) },
            { "swap", new FunctionBody(SWAP, null) },
            { "over", new FunctionBody(OVER, null) },
            { "+", new FunctionBody(ADD, null) },
            { "-", new FunctionBody(SUBTRACT, null) },
            { "*", new FunctionBody(MULTIPLY, null) },
            { "/", new FunctionBody(DIVIDE, null) },
            { ".", new FunctionBody(PRINT, null) },
        };

        bool EOF = false;
        while (!EOF)
        {

            string? instruction = Console.ReadLine();
            if (instruction == null)
                EOF = true;
            else
                ParseLine(instruction, stack, symbolTable);
        }
    }
*/
}

File: python/forth/forth.py
"""
Simple interpreter for a subset of Forth
"""
import re

class StackUnderflowError(Exception):
    """
    Custom Stack Underflow Error exception
    """
    pass


def evaluate(input_data):
    """Evaluate the given program
    - input_data: List of strings containing a forth program
    Returns: The stack (as a list)
    """
    if len(input_data) == 0:
        return [] # do nothing on empty program.

    stack = []
    symbols = [
        {'name': '+', 'value': add},
        {'name': '-', 'value': subtract},
        {'name': '*', 'value': multiply},
        {'name': '/', 'value': divide},
        {'name': 'dup', 'value': dup},
        {'name': 'drop', 'value': drop},
        {'name': 'swap', 'value': swap},
        {'name': 'over', 'value': over},
    ]
    for program in input_data:
        index = 0
        done = False 
        while not done:
            token = get_token(program, index)
            match token['type']:
                case 'Name':
                    index = token['end']
                    evaluate_token(stack, symbols, token)
                case 'Number':
                    index = token['end']
                    evaluate_token(stack, symbols, token)
                case 'BeginDefinition':
                    index = token['end']
                    index = parse_definition(program, symbols, index)
                case 'EndDefinition':
                    raise Exception('Unexpected end of definition')
                case 'EOF':
                    done = True
            if index > len(program):
                done = True
    return stack

def check_stack(stack, required_operands = 2):
    """ 
    Is the stack in the desired state for this function call. 
    - stack: current contents of the stack
    - required_operands: number of operands expected on the stack (default is 2)
    Returns True if sufficient operants, throws a StackUnderflowError if not enough found.
    """
    if len(stack) == 0:
        raise StackUnderflowError('Insufficient number of items in stack')
    
    if len(stack) < required_operands:
        raise StackUnderflowError('Insufficient number of items in stack')

    return True 

def evaluate_token(stack, symbols, token, lookback = None):
    """
    Evaluate a name or number token. (Just names and number so you can't nest a definition
    inside of a definition). EOF is handled in the main evaluate function.
    stack - current stack 
    symbols - symbol table 
    token - current token a dictionary of type, name, value, start and end
    lookback - where to start looking in the symbol table for a name
    """
    function = None
    match token['type']:
        case 'Number':
            stack.append(token['value'])
        case 'Name':
            function = find_name(symbols, token['value'], lookback)
            if function is None:
                raise ValueError('undefined operation')
            function(stack)


def evaluate_tokens(stack, symbols, tokens, lookback = None):
    """
    Evaluate a list of tokens
    stack - current stack 
    symbols - symbol table 
    tokens - a list of tokens to evaluate
    lookback - where to start looking in the symbol table for a name

    """
    for token in tokens:
        evaluate_token(stack, symbols, token, lookback)

def find_name(symbols, name, lookback = None):
    """ 
    find a name in the symbol table, starting at lookback/end of list 
    and iterating backwards to find it. Return null if not found.
    symbols - symbol table 
    name - name of the definition we are looking for
    lookback - where to start looking in the symbol table for a name

    """
    target = name.lower()
    begin = lookback                # len(symbols) - 1
    if begin is None:
        begin = len(symbols) - 1

    for index in range(begin, -1, -1):
        if target == symbols[index]['name']:
            return symbols[index]['value']
    return None


def get_token(text, start):
    """
    Get the next token from the given input text
    text - The string we are parsing
    start - offset into the string to parse from
    """
    re_ws = r'^\s+'                                     # white space
    re_num = r'^-?\d+'                                  # numbers (integer only)
    re_name = r'^([a-zA-Z][a-zA-Z0-9_-]*|\+|-|\*|\/)'   # names
    index = start 
    while index < len(text):
        current = text[index:]
        ret = re.search(re_ws, current)
        if ret is not None:
            index += ret.span()[1]
            continue

        ret = re.search(re_num, current)
        if ret is not None:
            num = int(ret.group())
            end_index = index + ret.span()[1]
            return { 
                'type': 'Number',
                'value': num,
                'start': index,
                'end': end_index,
            }    

        if text[index] == ':':
            return {
                'type': 'BeginDefinition',
                'value': ':',
                'start': index,
                'end': index + 1,
            }

        if text[index] == ';':
            return {
                'type': 'EndDefinition',
                'value': ';',
                'start': index,
                'end': index + 1,
            }

        ret = re.search(re_name, current)
        if ret is not None:
            end_index = index + ret.span()[1]
            return { 
                'type': 'Name',
                'value': ret.group(),
                'start': index,
                'end': end_index,
            }    

        raise Exception(f'Unrecognized token at {index}')

    return {
        'type': 'EOF',
        'value': '',
        'start': index,
        'end': index
    }

def parse_definition(program, symbols, index):
    """
    Parse a symbol definition
    program - the string we are parsing the program definition from
    index - offset into the program data
    """
    tokens = []
    position = index

    # expect name
    token = get_token(program, position)
    if token['type'] != 'Name':
        raise ValueError('illegal operation')
    name = token['value']
    position = token['end']

    while token['type'] != 'EndDefinition':    
        token = get_token(program, position)
        if token['type'] == 'EOF':
            raise Exception('Unexpected end of file')
        position = token['end']
        if token['type'] != 'EndDefinition':
            tokens.append(token)

    if token['type'] == 'EndDefinition':
        # lookback is to only look at definitions from when the word was defined.
        # you can get some nasty infinite loops otherwise.
        lookback = len(symbols) - 1

        # save the word
        # the function to call it encapsulates the symbol table, tokens, and lookback and just calls evaluateTokens on them.
        symbols.append({
            'name': name.lower(), 
            'value': lambda stack: evaluate_tokens(stack, symbols, tokens, lookback)
        })
    return position

# --------------------
# Built-in functions
# --------------------
def add(stack):
    """
    Add two numbers place the result onto the stack
    stack: we pull the numbers operate from off the stack
    """
    check_stack(stack)

    second = stack.pop()
    first = stack.pop()
    stack.append(first + second)


def subtract(stack):
    """
    Subtrack one number from another and place the result onto the stack
    stack: we pull the numbers operate from off the stack
    """
    check_stack(stack)

    second = stack.pop()
    first = stack.pop()
    stack.append(first - second)


def multiply(stack):
    """
    Multiply two numbers place the result onto the stack
    stack: we pull the numbers operate from off the stack
    """    
    check_stack(stack)

    second = stack.pop()
    first = stack.pop()
    stack.append(first * second)


def divide(stack):
    """
    Integer divide one number by another and place the result onto the stack
    stack: we pull the numbers operate from off the stack
    """
    check_stack(stack)

    second = stack.pop()
    first = stack.pop()
    if second == 0:
        raise ZeroDivisionError('divide by zero')
    stack.append(first // second)    # Change // to / for non-integer division


def dup(stack):
    """
    Duplicate the top item on the stack
    stack: The stack we operate on
    """    
    check_stack(stack, 1)
    first = stack.pop()

    stack.append(first)
    stack.append(first)

def drop(stack):
    """
    Drop the top item from the stack
    stack: The stack we operate on
    """    
    check_stack(stack, 1)
    stack.pop()

def swap(stack):
    """
    Swap the top two items on the stack
    stack: The stack we operate on
    """    
    check_stack(stack)
    second = stack.pop()
    first = stack.pop()

    stack.append(second)
    stack.append(first)


def over(stack):
    """
    Duplicate the second item in the stack to the top of the stack.
    stack: The stack we operate on
    """    
    check_stack(stack)
    second = stack.pop()
    first = stack.pop()

    stack.append(first)
    stack.append(second)
    stack.append(first)